File: app/src/main/java/calendario/kevshupp/diariokevinali/RecipeFragment.java
package calendario.kevshupp.diariokevinali;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;

import java.util.ArrayList;
import java.util.List;

public class RecipeFragment extends Fragment {

    private String coupleId, userId, userName, theme;
    private FirebaseFirestore db;
    private RecipeAdapter adapter;
    private List<Recipe> recipes = new ArrayList<>();
    private RecipeManager recipeManager; // Reutilizamos para los diálogos de añadir/editar

    public static RecipeFragment newInstance(String coupleId, String userId, String userName, String theme) {
        RecipeFragment fragment = new RecipeFragment();
        Bundle args = new Bundle();
        args.putString("coupleId", coupleId);
        args.putString("userId", userId);
        args.putString("userName", userName);
        args.putString("theme", theme);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            coupleId = getArguments().getString("coupleId");
            userId = getArguments().getString("userId");
            userName = getArguments().getString("userName");
            theme = getArguments().getString("theme");
        }
        db = FirebaseFirestore.getInstance();
        recipeManager = new RecipeManager(requireContext(), coupleId, userId, userName, () -> {
            if (getActivity() instanceof MainActivity) {
                ((MainActivity) getActivity()).pickImage(5); // PICK_IMAGE_RECIPE
            }
        });
        recipeManager.setTheme(theme);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_recipe_list, container, false);

        TextView tvTitle = view.findViewById(R.id.tvRecipeListTitle);
        TextView tvSubtitle = view.findViewById(R.id.tvRecipeListSubtitle);
        Button btnAdd = view.findViewById(R.id.btnAddRecipe);

        RecyclerView rvRecipes = view.findViewById(R.id.rvRecipes);
        adapter = new RecipeAdapter(recipes, userId, new RecipeAdapter.RecipeActionListener() {
            @Override public void onRecipeClick(Recipe recipe) { recipeManager.showRecipeDetail(recipe); }
            @Override public void onEditClick(Recipe recipe) { 
                recipeManager.showAddOrEditRecipeDialog(recipe);
            }
            @Override public void onDeleteClick(Recipe recipe) {
                db.collection("recipes").document(recipe.getRecipeId()).delete()
                    .addOnSuccessListener(aVoid -> Toast.makeText(getContext(), "Receta eliminada", Toast.LENGTH_SHORT).show());
            }
        });
        adapter.setTheme(theme);
        rvRecipes.setLayoutManager(new LinearLayoutManager(getContext()));
        rvRecipes.setAdapter(adapter);

        if ("Pixel Oscuro".equals(theme)) {
            view.setBackgroundResource(R.drawable.bg_parchment_pixel_dark);
            tvTitle.setTextColor(Color.WHITE);
            tvSubtitle.setTextColor(Color.LTGRAY);
            btnAdd.setTextColor(Color.WHITE);
            btnAdd.setBackgroundTintList(android.content.res.ColorStateList.valueOf(Color.parseColor("#1A1A2E")));
        } else {
            view.setBackgroundResource(R.drawable.bg_parchment_pixel);
        }

        listenRecipes();

        btnAdd.setOnClickListener(v -> recipeManager.showAddOrEditRecipeDialog(null));

        return view;
    }

    private void listenRecipes() {
        db.collection("recipes")
                .whereEqualTo("coupleId", coupleId)
                .addSnapshotListener((shots, error) -> {
                    if (shots == null) return;
                    recipes.clear();
                    for (QueryDocumentSnapshot doc : shots) {
                        Recipe recipe = doc.toObject(Recipe.class);
                        if (recipe.getRecipeId() == null) recipe.setRecipeId(doc.getId());
                        recipes.add(recipe);
                    }
                    recipes.sort((r1, r2) -> Long.compare(r2.getTimestamp(), r1.getTimestamp()));
                    adapter.notifyDataSetChanged();
                });
    }
    
    public void setImageUrl(String url) {
        if (recipeManager != null) recipeManager.setImageUrl(url);
    }
}

File: app/src/main/java/calendario/kevshupp/diariokevinali/RecipeFragment.kt
package calendario.kevshupp.diariokevinali

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.fragment.app.Fragment
import calendario.kevshupp.diariokevinali.compose.RecipeListScreen
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query

class RecipeFragment : Fragment() {
    private var coupleId: String? = null
    private var theme: String = "Pixel Claro"
    private var currentUserId: String? = null
    private val db = FirebaseFirestore.getInstance()
    private var recipeListener: ListenerRegistration? = null

    companion object {
        @JvmStatic
        fun newInstance(coupleId: String, theme: String): RecipeFragment {
            val f = RecipeFragment()
            val a = Bundle()
            a.putString("coupleId", coupleId)
            a.putString("theme", theme)
            f.arguments = a
            return f
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        coupleId = arguments?.getString("coupleId")
        theme = arguments?.getString("theme") ?: "Pixel Claro"
        val prefs = activity?.getSharedPreferences("DiarioPrefs", android.content.Context.MODE_PRIVATE)
        currentUserId = prefs?.getString("userId", null)
    }

    fun setImageUrl(url: String) {}

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return ComposeView(requireContext()).apply {
            setContent {
                var recipes by remember { mutableStateOf(listOf<Recipe>()) }
                val isDark = theme == "Pixel Oscuro"
                val activityRef = activity as? MainActivity

                LaunchedEffect(Unit) {
                    if (coupleId.isNullOrEmpty()) {
                        val prefs = activityRef?.getSharedPreferences("DiarioPrefs", android.content.Context.MODE_PRIVATE)
                        coupleId = prefs?.getString("coupleId", coupleId)
                        currentUserId = prefs?.getString("userId", currentUserId)
                        theme = prefs?.getString("theme", theme) ?: theme
                    }
                }

                DisposableEffect(coupleId) {
                    recipeListener?.remove()
                    recipeListener = db.collection("recipes")
                        .orderBy("timestamp", Query.Direction.DESCENDING)
                        .addSnapshotListener { value, _ ->
                            if (value != null) {
                                val wantedId = coupleId
                                val wantedIdNormalized = normalizeCoupleId(wantedId)
                                val ownerId = currentUserId
                                val filtered = value.mapNotNull { doc ->
                                    val recipe = doc.toObject(Recipe::class.java).apply { recipeId = doc.id }
                                    val docCoupleId = doc.getString("coupleId") ?: recipe.coupleId
                                    val docPartnerId = doc.getString("partnerId")
                                    val docAltId = doc.getString("couple_id")
                                    val docAuthorId = recipe.authorId ?: doc.getString("authorId")

                                    val matchesCouple = listOf(docCoupleId, docPartnerId, docAltId)
                                        .filterNotNull()
                                        .any { candidate ->
                                            candidate == wantedId || normalizeCoupleId(candidate) == wantedIdNormalized
                                        }

                                    val matchesOwner = ownerId != null && docAuthorId == ownerId
                                    if (matchesCouple || (docCoupleId == null && docPartnerId == null && docAltId == null && matchesOwner)) {
                                        recipe
                                    } else {
                                        null
                                    }
                                }
                                recipes = filtered
                            }
                        }
                    onDispose {
                        recipeListener?.remove()
                        recipeListener = null
                    }
                }

                RecipeListScreen(
                    recipes = recipes,
                    isDarkTheme = isDark,
                    onRecipeClick = { recipe -> activityRef?.openRecipeDetailDialog(recipe) },
                    onAddRecipeClick = { activityRef?.openAddRecipeDialog() }
                )
            }
        }
    }

    private fun normalizeCoupleId(value: String?): String? {
        return value
            ?.lowercase()
            ?.replace("á", "a")
            ?.replace("é", "e")
            ?.replace("í", "i")
            ?.replace("ó", "o")
            ?.replace("ú", "u")
            ?.replace("ñ", "n")
    }
}
